### PR TITLE
fix(scheduler_one): call Done() as soon as possible

### PR DESCRIPTION
**Pull request by sanposhiho**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

call Done() as soon as possible -- specifically after PreBind.

---

All cluster events during scheduling are piled up in [inFlightEvents](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/internal/queue/scheduling_queue.go#L187).
Basically, we need them to evaluate which queue to push failed Pods to. But, after https://github.com/kubernetes/kubernetes/pull/119105, we always push Pods to activeQ/backoffQ when those Pods are failed at extension points other than PreFilter, Filter, Reserve, WaitOnPermit, and PreBind.
In other words, if Pods go through PreBind, we don't need to have cluster events for those Pods. 
By calling Done() as soon as possible, we can reduce the memory consumed in inFlightEvents.

#### Which issue(s) this PR fixes:

Part of #120622

#### Does this PR introduce a user-facing change?

```release-note
When SchedulerQueueingHints is enabled, clear events cached in the scheduling queue as soon as possible so that the scheduler consumes less memory.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/scheduler/schedule_one.go [hunk 1]
```diff
@@ -283,6 +283,15 @@ func (sched *Scheduler) bindingCycle(
...
 	}
 
+	// All other extension points after here aren't supposed to fail in a usual state.
+	// i.e., They could return error only when some unexpected temporal errors happen. (e.g., kube-apiserver failure)
+	// So, if Pods are failed at one of them and going to the scheduling queue, the scheduling queue always pushes them to activeQ/backoffQ, instead of the unschedulable pod pool.
+	// This is because such Pods won't be schedulable by any certain cluster events, and thus scheduling hints aren't useful to determine when to push those Pods to activeQ/backoffQ.
+	//
+	// We can call Done() here because Pods failed at all other extension points after here won't use the scheduling hints as written.
+	// By calling Done(), we can free the cluster events stored in the scheduling queue, which is worth for busy clusters.
+	sched.SchedulingQueue.Done(assumedPod.UID)
```
Comment (sanposhiho) [inline pkg/scheduler/schedule_one.go:293]: If, for example, this Pod is failed at bind, Done() is called again when pushing this Pod to the scheduling queue (in AddUnschedulableIfNotPresent). But, that's fine because Done() is created to be safe to call many times.
https://github.com/sanposhiho/kubernetes/blob/02fa38a74aeb55e247bde3ee650fcd7b639a12ef/pkg/scheduler/internal/queue/scheduling_queue.go#L868

**Comment by sanposhiho** [comment]
Will check the test failure tomorrow.

File: pkg/scheduler/schedule_one.go [hunk 2]
```diff
@@ -283,6 +283,15 @@ func (sched *Scheduler) bindingCycle(
 		return status
 	}
 
+	// All other extension points after here aren't supposed to fail in a usual state.
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:286]:
```suggestion
	// Any failures after this point cannot lead to the Pod being considered unshcedulable.
```

File: pkg/scheduler/schedule_one.go [hunk 3]
```diff
@@ -283,6 +283,15 @@ func (sched *Scheduler) bindingCycle(
...
+	// i.e., They could return error only when some unexpected temporal errors happen. (e.g., kube-apiserver failure)
+	// So, if Pods are failed at one of them and going to the scheduling queue, the scheduling queue always pushes them to activeQ/backoffQ, instead of the unschedulable pod pool.
+	// This is because such Pods won't be schedulable by any certain cluster events, and thus scheduling hints aren't useful to determine when to push those Pods to activeQ/backoffQ.
+	//
+	// We can call Done() here because Pods failed at all other extension points after here won't use the scheduling hints as written.
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:291]:
```suggestion
	// We can call Done() here because Pods failed at all other extension points after here will be retried regardless of the events that happened in the cluster.
```

File: pkg/scheduler/schedule_one.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/scheduler/schedule_one.go:293]: I'm not sure if it's safe to call Done more than once.

Keep in mind that, because of status updates, a Pod could enter the scheduling cycle more than once, even if the scheduling cycle succeeded.

Comment (sanposhiho) [reply pkg/scheduler/schedule_one.go:293]: > a Pod could enter the scheduling cycle more than once

Oh, right, so the scenario like https://github.com/kubernetes/kubernetes/issues/118226.
OK, let's change here not to call Done() twice.

Comment (sanposhiho) [reply pkg/scheduler/schedule_one.go:293]: Yes, that could. I meant to have a temporal solution (like have a flag representing whether we've called Done() or not) though, after trying some, looks like it's going to be not a simple one.
Let's just wait for #119607 and merge this PR after that.

/hold

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 4]
```diff
@@ -616,6 +616,13 @@ func (p *PriorityQueue) SchedulingCycle() int64 {
 func (p *PriorityQueue) determineSchedulingHintForInFlightPod(logger klog.Logger, pInfo *framework.QueuedPodInfo) framework.QueueingHint {
 	logger.V(5).Info("Checking events for in-flight pod", "pod", klog.KObj(pInfo.Pod), "unschedulablePlugins", pInfo.UnschedulablePlugins, "inFlightEventsSize", p.inFlightEvents.Len(), "inFlightPodsSize", len(p.inFlightPods))
 
+	if len(pInfo.UnschedulablePlugins) == 0 {
```
Comment (sanposhiho) [inline pkg/scheduler/internal/queue/scheduling_queue.go:619]: Check UnschedulablePlugins before checking in-flight Pods.
It's necessary because we call Done() before `AddUnschedulableIfNotPresent` and `inFlightPod` wouldn't be found.

**Comment by alculquicondor** [comment]
Can you run some memory profiling based on this, perhaps using scheduling_perf?

**Comment by alculquicondor** [comment]
@ahg-g can take over the review from me if this requires extra changes on Friday or next week.

**Comment by alculquicondor** [comment]
@sanposhiho can you rebase this?
I would be great to merge something along these line before the code freeze.

We have an internal suite that we can use to validate whether it helps and report back.

**Comment by sanposhiho** [comment]
OK, will do tomorrow morning JST.

File: pkg/scheduler/schedule_one.go [hunk 5]
```diff
@@ -283,6 +283,15 @@ func (sched *Scheduler) bindingCycle(
 		return status
 	}
 
+	// Any failures after this point cannot lead to the Pod being considered unshcedulable.
+	// i.e., They could return error only when some unexpected temporal errors happen. (e.g., kube-apiserver failure)
+	// So, if Pods are failed at one of them and going to the scheduling queue, the scheduling queue always pushes them to activeQ/backoffQ, instead of the unschedulable pod pool.
+	// This is because such Pods won't be schedulable by any certain cluster events, and thus scheduling hints aren't useful to determine when to push those Pods to activeQ/backoffQ.
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/schedule_one.go:289]: On second thoughts, this PR can be merged without https://github.com/kubernetes/kubernetes/pull/119607.

Without https://github.com/kubernetes/kubernetes/pull/119607, this kind of scenario could happen:
1. Pod goes through the scheduling cycle successfully. The binding cycle is started at goroutineA.
2. [goroutineA] Done() is called right before binding. This Pod is removed from `inFlightPods`.
3. The Pod is updated by someone and, due to https://github.com/kubernetes/kubernetes/issues/118226, this Pod enters the scheduling cycle while the binding cycle of the same Pod is still on-going. The Pod is again added to `inFlightPods` in Pop().
4. [goroutineA] Binding fails. This Pod will be moved back to the scheduling queue.
5. [goroutineA] `AddUnschedulableIfNotPresent` is called to push this Pod into the scheduling queue. And Done() is again called in `AddUnschedulableIfNotPresent`. This Done() removes `inFlightPods` added at (3).
6. The scheduling cycle started at (3) is rejected by some plugins. This Pod is moved back to the scheduling queue. `AddUnschedulableIfNotPresent` is called again, but this Pod isn't registered in `inFlightPods` (due to (5)). 

But, (6) doesn't matter actually because the Pod that came back to the queue at (5) should be still in the queue, and this coming back Pod at (6) is just ignored ([ref](https://github.com/sanposhiho/kubernetes/blob/3e438c54a77ccf7b9e0aff924eb977031099cc2f/pkg/scheduler/internal/queue/scheduling_queue.go#L766-L775)) 

In the end, people would see an error log from the scheduler ([ref](https://github.com/sanposhiho/kubernetes/blob/3e438c54a77ccf7b9e0aff924eb977031099cc2f/pkg/scheduler/internal/queue/scheduling_queue.go#L766-L775)), but that's it. (Also, seemingly we can say this scenario is rare.)
I cannot imagine any worse scenario that could make Pod being stuck in the unschedQ by merging this PR without #119607.

**Comment by dims** [comment]
removing `needs-rebase` label by hand, automation did not clear it!

**Comment by sanposhiho** [comment]
/remove-lifecycle rotten
/reopen

This is still relevant.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 6]
```diff
@@ -654,6 +654,14 @@ func (p *PriorityQueue) SchedulingCycle() int64 {
 func (p *PriorityQueue) determineSchedulingHintForInFlightPod(logger klog.Logger, pInfo *framework.QueuedPodInfo) queueingStrategy {
 	logger.V(5).Info("Checking events for in-flight pod", "pod", klog.KObj(pInfo.Pod), "unschedulablePlugins", pInfo.UnschedulablePlugins, "inFlightEventsSize", p.inFlightEvents.Len(), "inFlightPodsSize", len(p.inFlightPods))
 
+	failedPlugins := pInfo.UnschedulablePlugins.Union(pInfo.PendingPlugins)
+	if len(failedPlugins) == 0 {
+		// No failed plugins are associated with this Pod.
+		// Meaning something unusual (a temporal failure on kube-apiserver, etc) happened and this Pod gets moved back to the queue.
+		// In this case, we should retry scheduling it because this Pod may not be retried until the next flush.
+		return queueAfterBackoff
+	}
```
Comment (ahg-g) [inline pkg/scheduler/internal/queue/scheduling_queue.go:663]: Curious why did you move this? this check is called at line 672 anyway. Just wondering if this change is based on something specific since it is not discussed in the issue description.

File: pkg/scheduler/schedule_one.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/schedule_one.go:289]: isn't done already called at handleSchedulingFailure, which gets invoked in case of a binding error? Why do we need it here?

https://github.com/kubernetes/kubernetes/blob/85d903e27ee92e999839e674577a3e79400835a0/pkg/scheduler/schedule_one.go#L985

[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/schedule_one.go:289]: Yes, it's already called, but the point is calling it before Bind, which contains API call, so that we can reduce the number of events stored in inFlightEvents (https://github.com/kubernetes/kubernetes/issues/120622).

[PR description:](https://github.com/kubernetes/kubernetes/pull/120586)
> All cluster events during scheduling are piled up in [inFlightEvents](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/internal/queue/scheduling_queue.go#L187).
...
By calling Done() as soon as possible, we can reduce the memory consumed in inFlightEvents.

File: pkg/scheduler/internal/queue/scheduling_queue.go (hunk 6 above)
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go:663]: The reason is described here: https://github.com/kubernetes/kubernetes/pull/120586#discussion_r1330260118.

By the change in pkg/scheduler/schedule_one.go, if a Pod is rejected at the bind extension point, `p.inFlightPods[pInfo.Pod.UID]` would be not-found (as `Done()` removed the Pod from `p.inFlightPods` before calling the bind extension point). That'd be a problem of the original `determineSchedulingHintForInFlightPod`.

Actually, Pods rejected at the binding extension point should always go to backoffQ and we actually don't have to look up `p.inFlightPods`. 
So, here we can just reverse the order of looking up `p.inFlightPods` and checking `failedPlugins` so that such bind-failure scenario is also covered correctly.

**Comment by dims** [comment]
@sanposhiho  please rebase!

**Comment by sanposhiho** [comment]
Rebased, now https://github.com/kubernetes/kubernetes/issues/118226 got solved and we're ready to merge it with no concern at all.

@kubernetes/sig-scheduling-approvers can anyone take a look for the approve?

**Comment by sanposhiho** [comment]
cc @macsko 

Can you possibly take a look? Sorry for having mentioned you my PRs everywhere 😅

File: pkg/scheduler/schedule_one.go [hunk 7]
```diff
@@ -306,6 +306,13 @@ func (sched *Scheduler) bindingCycle(
 		return status
 	}
 
+	// Any failures after this point cannot lead to the Pod being considered unschedulable.
+	// We define the Pod as "unschedulable" only when Pods are rejected at specific extension points, and PreBind is the last one in the scheduling/binding cycle.
+	//
+	// We can call Done() here because
+	// we can free the cluster events stored in the scheduling queue sonner, which is worth for busy clusters memory consumption wise.
+	sched.SchedulingQueue.Done(assumedPod.UID)
```
Comment (macsko) [inline pkg/scheduler/schedule_one.go:314]: Shouldn't `Done` call be removed from [there](https://github.com/kubernetes/kubernetes/blob/01eb9f47541156a38e4b28759977c06f07f08012/pkg/scheduler/schedule_one.go#L132)?

Comment (sanposhiho) [reply pkg/scheduler/schedule_one.go:311]: Correct, that's a reductant Done.
Removed.

**Comment by sanposhiho** [comment]
cc @alculquicondor 
for `/approve`

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
When SchedulerQueueingHints is enabled, clear events cached in the scheduling queue as soon as possible so that the scheduler consumes less memory.
```

**Review by alculquicondor — COMMENTED** [review]
/approve
/hold
to clarify the comments

File: pkg/scheduler/framework/types.go [hunk 8]
```diff
@@ -218,8 +218,12 @@ type QueuedPodInfo struct {
 	// It shouldn't be updated once initialized. It's used to record the e2e scheduling
 	// latency for a pod.
 	InitialAttemptTimestamp *time.Time
-	// UnschedulablePlugins records the plugin names that the Pod failed with Unschedulable or UnschedulableAndUnresolvable status.
-	// It's registered only when the Pod is rejected in PreFilter, Filter, Reserve, PreBind or Permit (WaitOnPermit).
+	// UnschedulablePlugins records the plugin names that the Pod failed with Unschedulable or UnschedulableAndUnresolvable status
+	// at specific extension points: PreFilter, Filter, Reserve, Permit (WaitOnPermit), or PreBind.
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/framework/types.go:222]: We added PreBind here after I started this PR and I just forgot updating the description.

File: pkg/scheduler/internal/queue/scheduling_queue.go [hunk 9]
```diff
@@ -738,12 +738,6 @@ func (p *PriorityQueue) clusterEventsForPod(logger klog.Logger, pInfo *framework
...
 // and determines the scheduling hint for this Pod while checking the events that happened during in-flight.
 func (p *PriorityQueue) determineSchedulingHintForInFlightPod(logger klog.Logger, pInfo *framework.QueuedPodInfo) queueingStrategy {
-	events, err := p.clusterEventsForPod(logger, pInfo)
-	if err != nil {
-		logger.Error(err, "Error getting cluster events for pod", "pod", klog.KObj(pInfo.Pod))
-		return queueAfterBackoff
-	}
-
 	rejectorPlugins := pInfo.UnschedulablePlugins.Union(pInfo.PendingPlugins)
 	if len(rejectorPlugins) == 0 {
```
[the remark this answers is not in the document]
Comment (sanposhiho) [reply pkg/scheduler/internal/queue/scheduling_queue.go]: I included in this PR.

**Comment by sanposhiho** [comment]
@alculquicondor Addressed your points.

**Review by alculquicondor — COMMENTED** [review]
/lgtm

**Comment by alculquicondor** [comment]
/retest
Opened #126950 for the flaky
